File: web/src/api/dto/imports.ts
import type { Id } from '../types'
import type { AccountDto } from './account'
import type { CreateTransactionDto, TransactionDto, TransactionType } from './transaction'

export type ImportProvider = 'apple-wallet' | 'simplefin'
export type ImportCardState = 'mapped' | 'ignored' | 'unmapped'
export type ImportQueueReason = 'unmapped' | 'account_deleted' | 'no_rate'
export type IngestStatus = 'created' | 'queued' | 'skipped' | 'duplicate' | 'failed'
export type ImportRunStatus = 'running' | 'completed' | 'partial' | 'failed'

export interface ImportCardDto {
  externalAccountId: string
  externalName: string
  externalCurrency: string
  state: ImportCardState
  /** '' unless state === 'mapped' */
  accountId: Id | ''
  queuedCount: number
  tapCount: number
  /** "YYYY-MM-DD HH:mm:ss" of the newest event from this card, '' when none */
  lastSeenAt: string
}

export interface ImportSourceDto {
  id: Id
  provider: ImportProvider
  name: string
  status: string
  createdAt: string
  /** "YYYY-MM-DD HH:mm:ss" of the last non-failed sync, '' before the first */
  lastSyncedAt: string
  /** opaque to the server; '' for push providers */
  credentialCiphertext: string
  cards: ImportCardDto[]
}

export interface ImportRunErrorDto {
  /** '' for a run-level (bridge) error */
  externalAccountId: string
  message: string
}

export interface ImportRunDto {
  id: Id
  sourceId: Id
  provider: ImportProvider
  status: ImportRunStatus
  trigger: string
  importedCount: number
  matchedCount: number
  amountsUpdatedCount: number
  queuedCount: number
  skippedCount: number
  failedCount: number
  errors: ImportRunErrorDto[]
  startedAt: string
  /** '' while running */
  finishedAt: string
}

export interface ImportRunLinkDto {
  id: Id
  externalAccountId: string
  externalTransactionId: string
  /** '' for queued/skipped rows and for tombstones (transaction deleted since) */
  transactionId: Id | ''
  status: string
  externalPayee: string
  externalAmount: string
  externalCurrency: string
  externalPostedAt: string
}

export interface ImportRunDetailDto {
  item: ImportRunDto
  links: ImportRunLinkDto[]
}

export interface ExternalAccountDto {
  externalAccountId: string
  externalName: string
  externalCurrency: string
  balance: string
  orgName: string
  state: ImportCardState
  accountId: Id | ''
}

export interface ImportCredentialKeyDto {
  wrappedDataKey: string
  kdf: string
  updatedAt: string
}

export interface SyncImportSourceResultDto {
  run: ImportRunDto
  accounts: ExternalAccountDto[]
}

/** link-account / ignore-account / unlink-account: the refreshed source, plus
    the conversion run when link-account drained a queue (null otherwise). */
export interface UpdateImportAccountDto {
  item: ImportSourceDto
  run: ImportRunDto | null
}

export interface IngestEventDto {
  status: IngestStatus
  /** '' for a duplicate (nothing new was stored) */
  eventId: Id | ''
}

export interface ImportQueuedEventDto {
  linkId: Id
  sourceId: Id
  externalAccountId: string
  /** '' while the card is unmapped */
  accountId: Id | ''
  payee: string
  amount: string
  currency: string
  type: TransactionType
  /** "YYYY-MM-DD HH:mm:ss" */
  postedAt: string
  reason: ImportQueueReason
}

export interface ImportFailedEventDto {
  eventId: Id
  sourceId: Id
  receivedAt: string
  error: string
  /** the raw body as received, for the "needs attention" panel */
  payload: string
}

export interface ImportQueueDto {
  queued: ImportQueuedEventDto[]
  skipped: ImportQueuedEventDto[]
  failed: ImportFailedEventDto[]
}

export interface ImportQueuedEventPayload {
  linkId: Id
  transaction: CreateTransactionDto
}

export interface ImportQueuedEventResultDto {
  item: TransactionDto
  accounts: AccountDto[]
}

export interface TransactionImportLinkDto {
  id: Id
  sourceId: Id
  provider: ImportProvider
  sourceName: string
  externalAccountId: string
  externalTransactionId: string
  externalPayee: string
  externalAmount: string
  externalCurrency: string
  externalPostedAt: string
  status: string
  importedAt: string
}

File: internal/model/imports.go
package model

import (
	"time"

	"github.com/econumo/econumo/internal/shared/vo"
)

const (
	ImportProviderSimpleFIN   = "simplefin"
	ImportProviderCSV         = "csv"
	ImportProviderAppleWallet = "apple-wallet"

	ImportSourceStatusActive = "active"

	ImportEventStatusProcessed = "processed"
	ImportEventStatusFailed    = "failed"
	ImportEventStatusDiscarded = "discarded"

	ImportLinkStatusLinked  = "linked"
	ImportLinkStatusQueued  = "queued"
	ImportLinkStatusSkipped = "skipped"

	ImportRunStatusRunning   = "running"
	ImportRunStatusCompleted = "completed"
	ImportRunStatusFailed    = "failed"
	ImportRunStatusPartial   = "partial"

	ImportAccountLinkModeImport = "import"
	ImportAccountLinkModeIgnore = "ignore"
)

// Push providers deliver one event per tap and see the transaction before
// the bank posts it; pull providers see the posted record. The matcher
// treats the two differently (a push link is the "tip" a later bank record
// adopts).
func ImportProviderIsPush(provider string) bool {
	return provider == ImportProviderAppleWallet
}

type ImportSource struct {
	ID                   vo.Id
	UserID               vo.Id
	Provider             string
	Name                 string
	CredentialCiphertext *string
	Status               string
	LastSyncedAt         *time.Time
	CreatedAt            time.Time
	UpdatedAt            time.Time
}

// ImportAccountLink is the user's decision for one external account (a
// card name for Apple Wallet): import into AccountID, or ignore. AccountID
// is nil only in ignore mode (schema CHECK). ExternalCurrency is the
// currency the provider reported for the card, recorded when the link is
// created so a currency drift can be refused later.
type ImportAccountLink struct {
	ID                vo.Id
	SourceID          vo.Id
	ExternalAccountID string
	ExternalName      string
	ExternalCurrency  *string
	AccountID         *vo.Id
	Mode              string
	CreatedAt         time.Time
	UpdatedAt         time.Time
}

type ImportEvent struct {
	ID          vo.Id
	SourceID    vo.Id
	RunID       *vo.Id
	Payload     string
	PayloadHash string
	Status      string
	ParseError  *string
	ReceivedAt  time.Time
}

type ImportRun struct {
	ID            vo.Id
	UserID        vo.Id
	SourceID      vo.Id
	Provider      string
	Params        string
	Status        string
	ImportedCount int
	MatchedCount  int
	SkippedCount  int
	FailedCount   int
	StartedAt     time.Time
	FinishedAt    *time.Time
}

// ImportTransactionLink is one ledger row: what an external transaction id
// became. A linked row whose TransactionID is nil is a tombstone — the user
// deleted the transaction, and the external id must never be re-imported.
type ImportTransactionLink struct {
	ID                    vo.Id
	SourceID              vo.Id
	RunID                 *vo.Id
	EventID               *vo.Id
	ExternalAccountID     string
	ExternalTransactionID string
	TransactionID         *vo.Id
	Status                string
	ExternalPayee         string
	ExternalDescription   string
	ExternalAmount        string
	ExternalCurrency      *string
	ExternalPostedAt      time.Time
	AppliedCategoryID     *vo.Id
	AppliedPayeeID        *vo.Id
	AppliedTagID          *vo.Id
	AppliedRuleID         *vo.Id
	ImportedAt            time.Time
}

// A queued row is not "seen": the user has not decided yet, so the same
// external id may be re-offered by a later sync.
func (l ImportTransactionLink) IsSeen() bool { return l.Status != ImportLinkStatusQueued }

func (l ImportTransactionLink) IsTombstone() bool {
	return l.Status == ImportLinkStatusLinked && l.TransactionID == nil
}

// IngestEvent is one normalized external transaction, the matcher's input.
// Amount is the absolute value as a decimal string; Type carries the sign.
type IngestEvent struct {
	ExternalAccountID     string
	ExternalTransactionID string
	Type                  TransactionType
	Amount                string
	Currency              string
	PostedAt              time.Time
	Payee                 string
	Description           string
}

// ImportCandidate is an existing transaction the matcher may adopt, with the
// ledger rows already pointing at it.
type ImportCandidate struct {
	TransactionID vo.Id
	Type          TransactionType
	Amount        string
	SpentAt       time.Time
	Links         []ImportCandidateLink
}

type ImportCandidateLink struct {
	SourceID       vo.Id
	Provider       string
	ExternalAmount string
	ExternalPayee  string
}
